compiler/frontend/src/mir_optimize/pure.rs

```diff
@@ -1,114 +1,324 @@
+        }
+    }
+    pub fn is_function_deterministic(&self, expression: &Expression) -> bool {
+        match expression {
+            Expression::Builtin(builtin) => match builtin {
+                BuiltinFunction::Equals
+                | BuiltinFunction::GetArgumentCount
+                | BuiltinFunction::IntAdd
+                | BuiltinFunction::IntBitLength
+                | BuiltinFunction::IntBitwiseAnd
+                | BuiltinFunction::IntBitwiseOr
+                | BuiltinFunction::IntBitwiseXor
+                | BuiltinFunction::IntCompareTo
+                | BuiltinFunction::IntDivideTruncating
+                | BuiltinFunction::IntModulo
+                | BuiltinFunction::IntMultiply
+                | BuiltinFunction::IntParse
+                | BuiltinFunction::IntRemainder
+                | BuiltinFunction::IntShiftLeft
+                | BuiltinFunction::IntShiftRight
+                | BuiltinFunction::IntSubtract
+                | BuiltinFunction::ListFilled
+                | BuiltinFunction::ListGet
+                | BuiltinFunction::ListInsert
+                | BuiltinFunction::ListLength
+                | BuiltinFunction::ListRemoveAt
+                | BuiltinFunction::ListReplace
+                | BuiltinFunction::StructGet
+                | BuiltinFunction::StructGetKeys
+                | BuiltinFunction::StructHasKey
+                | BuiltinFunction::TagGetValue
+                | BuiltinFunction::TagHasValue
+                | BuiltinFunction::TagWithoutValue
+                | BuiltinFunction::TextCharacters
+                | BuiltinFunction::TextConcatenate
+                | BuiltinFunction::TextContains
+                | BuiltinFunction::TextEndsWith
+                | BuiltinFunction::TextFromUtf8
+                | BuiltinFunction::TextGetRange
+                | BuiltinFunction::TextIsEmpty
+                | BuiltinFunction::TextLength
+                | BuiltinFunction::TextStartsWith
+                | BuiltinFunction::TextTrimEnd
+                | BuiltinFunction::TextTrimStart
+                | BuiltinFunction::ToDebugText
+                | BuiltinFunction::TypeOf => true,
+                BuiltinFunction::FunctionRun | BuiltinFunction::IfElse | BuiltinFunction::Print => {
+                    false
+                }
+            },
+            Expression::Function { body, .. } => body
+                .iter()
+                .all(|(_, expression)| self.is_definition_deterministic(expression)),
+            Expression::Tag { .. } => true, // either works or panics
+            Expression::Int(_)
+            | Expression::Text(_)
+            | Expression::List(_)
+            | Expression::Struct(_)
+            | Expression::HirId(_)
+            | Expression::UseModule { .. } => true, // always panics
+            Expression::Parameter
+            | Expression::Call { .. }
+            | Expression::Panic { .. }
```

Comment:
I think panics could be counted as deterministic functions as well. When trying to call the result of a panic expression, the program will have panicked already, so we can deduplicate the code after it. OTOH, we have the new optimization `remove_expressions_after_panic`, so it doesn't really matter, but maybe add a comment.